### PR TITLE
Enable Fastboot with USE_FASTBOOT environment variable set to 1

### PR DESCRIPTION
**Pull request by kzys**
This PR enables Fastboot on `/policies` like before, but only if USE_FASTBOOT is defined.

### CODE REVIEW

File: config/nginx.conf.erb [hunk 1]
```diff
@@ -64,6 +64,11 @@ http {
 			proxy_pass http://app_server;
 		}
 
+		# Just in case, only forward "/policies" to Ember for a moment
+		location = /policies {
+			proxy_pass http://localhost:9000;
+		}
```
Comment (jtgeibel) [inline config/nginx.conf.erb]: The template should test `ENV['USE_FASTBOOT']` here, right?

Comment (kzys) [reply config/nginx.conf.erb]: Right! Not so sure how did I miss it...

**Comment by kzys** [comment]
Travis seems having a bad day.

File: script/start-web.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#! /bin/sh
+set -ue
+
+if [ "${USE_FASTBOOT:-0}" = 1 ]; then
+    export USE_FASTBOOT=1
+    node --optimize_for_size --max_old_space_size=200 fastboot.js &
+    bin/start-nginx ./target/release/server &
+    wait -n
```
Comment (kzys) [inline script/start-web.sh]: `wait -n` doesn't work on /bin/sh. It is only available on bash.

**Comment by carols10cents** [comment]
I have this deployed to staging right now with USE_FASTBOOT=TRUE, `/policies` is working fine!

I don't see any difference in the logs really, am I supposed to? Here's what I see in the logs for a request to `/policies`:

<details>

```
2019-11-15T14:39:48.323842+00:00 heroku[router]: at=info method=GET path="/policies" host=staging-crates-io.herokuapp.com request_id=e99bc704-d296-4755-a20e-e27de1ab3143 fwd="71.112.158.75,70.132.60.136" dyno=web.1 connect=0ms service=16ms status=200 bytes=2333 protocol=https
2019-11-15T14:39:48.320751+00:00 app[web.1]: at=info method=GET path="/policies" request_id=e99bc704-d296-4755-a20e-e27de1ab3143 fwd="70.132.60.136" service=0ms status=200 user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36"
2019-11-15T14:39:48.324643+00:00 app[web.1]: measure#nginx.service=0.014 request_id=e99bc704-d296-4755-a20e-e27de1ab3143
2019-11-15T14:39:48.629354+00:00 heroku[router]: at=info method=GET path="/cargo-835dd6a18132048a52ac569f2615b59d.png" host=staging-crates-io.herokuapp.com request_id=7130b9fa-b2ec-4506-b347-012d30e67e1d fwd="71.112.158.75,70.132.60.134" dyno=web.1 connect=0ms service=5ms status=200 bytes=60779 protocol=https
2019-11-15T14:39:48.625404+00:00 app[web.1]: at=info method=GET path="/cargo-835dd6a18132048a52ac569f2615b59d.png" request_id=7130b9fa-b2ec-4506-b347-012d30e67e1d fwd="70.132.60.134" service=0ms status=200 user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36"
2019-11-15T14:39:48.629663+00:00 app[web.1]: measure#nginx.service=0.006 request_id=7130b9fa-b2ec-4506-b347-012d30e67e1d
```

</details>

And here's what I see in the logs for a request to `/category_slugs`:

<details>

```
2019-11-15T14:39:13.606528+00:00 heroku[router]: at=info method=GET path="/category_slugs" host=staging-crates-io.herokuapp.com request_id=cbe927ce-c63c-4d8a-ae86-97a795b8a719 fwd="71.112.158.75,70.132.60.156" dyno=web.1 connect=0ms service=8ms status=200 bytes=2333 protocol=https
2019-11-15T14:39:13.603467+00:00 app[web.1]: at=info method=GET path="/category_slugs" request_id=cbe927ce-c63c-4d8a-ae86-97a795b8a719 fwd="70.132.60.156" service=0ms status=200 user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36"
2019-11-15T14:39:13.605816+00:00 app[web.1]: measure#nginx.service=0.005 request_id=cbe927ce-c63c-4d8a-ae86-97a795b8a719
2019-11-15T14:39:13.742656+00:00 app[web.1]: at=info method=GET path="/api/v1/category_slugs" request_id=7f31c9af-ee51-453d-84d3-1f9a4b7313b8 fwd="70.132.60.131" service=4ms status=200 user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36"
2019-11-15T14:39:13.749201+00:00 app[web.1]: measure#nginx.service=0.014 request_id=7f31c9af-ee51-453d-84d3-1f9a4b7313b8
2019-11-15T14:39:13.749153+00:00 heroku[router]: at=info method=GET path="/api/v1/category_slugs" host=staging-crates-io.herokuapp.com request_id=7f31c9af-ee51-453d-84d3-1f9a4b7313b8 fwd="71.112.158.75,70.132.60.131" dyno=web.1 connect=3ms service=16ms status=200 bytes=4003 protocol=https
```

</details>

I'm not seeing the memory issues reported in https://github.com/rust-lang/crates.io/pull/1827; in fact, the memory usage is lower after deploying this (previously staging had 77856b0f on it). I made sure to request `/policies` repeatedly, and I'm not seeing any memory usage growth:

staging metrics graphs 1
staging metrics graphs 2

**Comment by kzys** [comment]
Thanks @carols10cents. Have you deployed something else on staging-crates-io.herokuapp.com? https://staging-crates-io.herokuapp.com/policies doesn't seem to have Fastboot.

Regarding logs, you should at least have logs from https://github.com/ember-fastboot/fastboot-express-middleware/blob/3febb1dbc11b4b95e2125c6866de2498e9d4d08c/src/index.js#L91, which I couldn't find a way to disable that from fastboot-app-server.

**Comment by carols10cents** [comment]
I have not... [it's still at 6e373b6](https://staging-crates-io.herokuapp.com/api/v1/site_metadata). I set `USE_FASTBOOT` to `true` and just confirmed that it's still set, is that not right?

How can you tell if a page uses fastboot or not?

File: script/ember.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#! /bin/sh
+set -ue
+
+export FASTBOOT_DISABLED
+
+if [ "${USE_FASTBOOT:-0}" = '1' ]; then
```
Comment (carols10cents) [inline script/ember.sh]: oh oops it needs to be `1` here? let me try that

File: config/nginx.conf.erb [hunk 4]
```diff
@@ -64,6 +64,13 @@ http {
 			proxy_pass http://app_server;
 		}
 
+		<% if ENV['USE_FASTBOOT'] %>
```
Comment (carols10cents) [inline config/nginx.conf.erb]: i saw this line and thought it just needed to be set.

File: script/ember.sh (hunk 3 above)
Comment (kzys) [reply script/ember.sh]: ember.sh is only used from npm run, but start-web.sh is doing the same check. Yes. It has to be 1

**Comment by jtgeibel** [comment]
I've pushed a commit to [my staging](https://jtgeibel-staging-crates-io.herokuapp.com/) area that allows me to experimentally enable FastBoot for all front-end paths (via `USE_FASTBOOT=staging-experimental`).  This should make it easier to test things going forward.  We could include it in this PR, or I can wait and open a separate PR for this.

Branch `pr/1900-experimental`: https://github.com/jtgeibel/crates.io/tree/pr/1900-experimental
Diff: https://github.com/jtgeibel/crates.io/commit/4c33f00447a3ced0384122d84bb2e03849a51e6d

**Comment by carols10cents** [comment]
OK *now* I am seeing this in the logs for a request to `/policies`:

<details>

```
2019-11-15T19:16:39.834591+00:00 heroku[router]: at=info method=GET path="/policies" host=staging-crates-io.herokuapp.com request_id=12a23729-d07c-442e-953d-fac4fdbe1187 fwd="71.112.158.75" dyno=web.1 connect=1ms service=416ms status=200 bytes=5826 protocol=https
2019-11-15T19:16:39.826593+00:00 app[web.1]: 2019-11-15T19:16:39.826Z 200 OK /policies
2019-11-15T19:16:39.835736+00:00 app[web.1]: measure#nginx.service=0.410 request_id=12a23729-d07c-442e-953d-fac4fdbe1187
```

</details>

**Comment by jtgeibel** [comment]
Regarding the log format, my only suggestions would be to see if we can remove the extra timestamp from line 2, replacing it with "FastBoot:".  It would also be nice to log the request_id on that line.

```text
2019-11-15T19:16:39.826593+00:00 app[web.1]: FastBoot: 200 OK /policies request_id=12a23729-d07c-442e-953d-fac4fdbe1187
```

**Comment by carols10cents** [comment]
Ok I am seeing higher memory usage now, but not outrageously so. There was a jump from 37MB to 57MB when the dynos were done restarting after I fixed the env var value, and it fluctuated as high as 86MB, but I made a point to reload `/policies` a whole bunch and I didn't see any resulting change in memory usage. Here's the metrics graph:

fastboot metrics

**Comment by sgrif** [comment]
I'm seeing 500s for that page on staging

**Comment by sgrif** [comment]
Scratch that, was able to successfully access from another domain. I did a quick stress test sending 1k requests to `/policies`. Memory usage did increase for a bit, but topped out at ~30MB total which seems fine. Response times were slower than I'd like but staging isn't representative of production for that.

**Comment by kzys** [comment]
@jtgeibel Regarding the log message, changing that is hard because we cannot configure the logger through fastboot-app-server. I would like to send a PR for one of the Fastboot packages but it may take time.

Adding a new log to indicate Fastboot access is much easier. I can do that probably in this PR if you want to.

**Comment by jtgeibel** [comment]
@kzys I think the log format is fine for now, especially for this single endpoint.  There was already a related todo item on #1811 and I've linked to the comment above from there.  I think we'll want to eventually add the request_id to the output before adding too many more endpoints, but since that requires upstream changes I don't think that needs to block this particular PR.

I've assigned #1811 to you, in the hopes that it will enable you to edit that issue.  If not, I'll try to keep it synced with our progress as things land (I believe I've addressed your other comments in that thread).

**Comment by carols10cents** [comment]
Sounds like we're all good with this. I'm going to merge now, but not deploy til Wednesday when we're sure the cloudfront changes have gone well.

@bors r+

**Pull request merged**